backfill.sh

```diff
@@ -0,0 +1,140 @@
+#!/bin/bash
+# this script is used to update the article-json stored in lax
+# performing just an INGEST of content, *no* PUBLISH events are
+# sent whatsoever.
+
+# talking to lax via the adaptor.py script for many thousands of articles is
+# extremely slow, so this script bypasses all that, bulk generates articles,
+# bulk validates them then tells lax to do a bulk ingest.
+
+set -euo pipefail # strict mode
+#set -xv # debugging
+
+# housekeeping
+
+trap ctrl_c INT
+function ctrl_c() {
+    echo "caught ctrl-c"
+    exit 1
+}
+
+srcdir=$(pwd)
+errcho(){ >&2 echo $@; }
+
+# where articles will be linked to/downloaded for backfill
+defaultrunpath="run-$(date +'%Y%m%d%H%M%S')"
+runpath=${1:-$defaultrunpath}
+runpath=$(realpath $runpath) # realpath needs to be installed :(
+mkdir -p "$runpath"
+
+shift # download-api-raml.sh takes an arg :P
+
+# confirm
+
+echo "backfill.sh
+
+this script will:
+1. pull latest article-xml from elifesciences/elife-article-xml (large repo)
+2. download any missing/unpublished articles after consulting Lax (needs /srv/lax, s3 auth)
+3. create a 'backfill-run' directory with symbolic links to the xml to be processed
+4. generate article-json from ALL xml in the ./articles-xml/articles/ directory (long process)
+5. validate all generated article-json, failing if any are invalid
+6. force an INGEST into Lax for all valid articles (needs /srv/lax)"
+
+read -p "any key to continue (ctrl-c to quit) "
+
+# begin
+
+. download-elife-xml.sh
+
+# activate venv
+set +o nounset; . install.sh; set -o nounset; # virtualenv has unset vars we can't control
+
+# where to find xml on the fs
+xmlrepodir="$srcdir/article-xml/articles"
+
+# where to download unpublished xml to
+unpubxmldir="$srcdir/unpub-article-xml"
+mkdir -p "$unpubxmldir" # (create if necessary)
+
+# where generated article-json will be stored
+ajsondir="$runpath/ajson"
```

Comment:
as I was picking the code apart I could see we're actually using twice the amount of space than we need to. way back in the beginning we were 'patching' the generated article json to fill in the missing gaps and getting it passing validation so lax wouldn't reject it, lax would then fix up anything it needed to under the hood. 

the `generate-article-json` stage creates the article-json without the patched values and then the validate stage patches this article-json before validating it.

the backfill process is generating the article-json, then validating it and storing the patched version in `$ajsondir/patched` and then sending the patched version to lax.

I could halve the space requirements by adding the patching to `main.py` in the post-process step after scraping and having validate.py deal with just `valid` and `invalid`.

It wouldn't remove the need for an external disk - the article xml will grow over time - but it would make things more efficient and less unwieldly. I'm sure those inodes would thank me if they could as well.